src/main/java/org/icgc/argo/program_service/services/ProgramService.java

```diff
@@ -18,207 +18,311 @@
 
-    List<ProgramCancer> programCancers = mapToList(cancers, x -> createProgramCancer(program, x).get());
-    List<ProgramPrimarySite> programPrimarySites = mapToList(primarySites, x -> createProgramPrimarySite(program, x).get());
     program.setProgramCancers(new TreeSet<>(programCancers));
     program.setProgramPrimarySites(new TreeSet<>(programPrimarySites));
+    program.setProgramInstitutions(new TreeSet<>(programInstitutions));
+    program.setProgramRegions(new TreeSet<>(programRegions));
+    program.setProgramCountries(new TreeSet<>(programCountries));
 
     return program;
   }
 
-  //TODO: add existence check, and ensure program doesnt already exist. If it does, return a Conflict
+  @Transactional
+  public ProgramEntity createWithSideEffectTransactional(@NonNull Program program, Consumer<ProgramEntity> consumer) {
+    val programEntity = createProgram(program);
+    consumer.accept(programEntity);
+    return programEntity;
+  }
+
+  // TODO: add existence check, and ensure program doesnt already exist. If it does, return a
+  // Conflict
   public ProgramEntity createProgram(@NonNull Program program)
-    throws DataIntegrityViolationException {
+      throws DataIntegrityViolationException {
     val programEntity = programConverter.programToProgramEntity(program);
-    // Set the timestamps
     val now = LocalDateTime.now(ZoneId.of("UTC"));
     programEntity.setCreatedAt(now);
     programEntity.setUpdatedAt(now);
 
-    val p = programRepository.save(programEntity);
+    if (program.getCancerTypesList().isEmpty()
+        || program.getPrimarySitesList().isEmpty()
+        || program.getInstitutionsList().isEmpty()
+        || program.getRegionsList().isEmpty()
+        || program.getCountriesList().isEmpty()) {
+      throw Status.INVALID_ARGUMENT
+          .augmentDescription(
+            "Cannot create program, Must provide at least one of each: cancer, primary site, institution, country, and region.")
+          .asRuntimeException();
+    }
 
     val cancers = cancerRepository.findAllByNameIn(program.getCancerTypesList());
     val primarySites = primarySiteRepository.findAllByNameIn(program.getPrimarySitesList());
+    val institutions = institutionRepository.findAllByNameIn(program.getInstitutionsList());
+    val regions = regionRepository.findAllByNameIn(program.getRegionsList());
+    val countries = countryRepository.findAllByNameIn(program.getCountriesList());
+
+    if (cancers.size() != program.getCancerTypesList().size()
+      || primarySites.size() != program.getPrimarySitesList().size()
+      || institutions.size() != program.getInstitutionsList().size()
+      || countries.size() != program.getCountriesList().size()
+      || regions.size() != program.getRegionsList().size()) {
+      throw Status.INVALID_ARGUMENT
+        .augmentDescription(
+          "Cannot create program, Must provide valid cancer, primary site, institution, country, and region.")
```

Comment:
It is possible to write in a new institution name, I believe. Doesn't have to exist, can be created by this method.